contracts/lper/src/contract.rs

```diff
@@ -0,0 +1,166 @@
+};
+use cw2::set_contract_version;
+
+use astroport::{pair::{ExecuteMsg::ProvideLiquidity, Cw20HookMsg}, asset::{Asset, AssetInfo}};
+use cw20::Cw20ReceiveMsg;
+
+use crate::{msg::{ExecuteMsg, InstantiateMsg, MigrateMsg, QueryMsg}, state::{HOLDER_ADDRESS, LP_POSITION}};
+
+use neutron_sdk::{
+    bindings::{
+        msg::{NeutronMsg},
+        query::{NeutronQuery},
+    },
+    NeutronResult,
+};
+
+use crate::state::{
+   CLOCK_ADDRESS, CONTRACT_STATE, ContractState,
+};
+
+
+const CONTRACT_NAME: &str = "crates.io:covenant-lper";
+const CONTRACT_VERSION: &str = env!("CARGO_PKG_VERSION");
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn instantiate(
+    deps: DepsMut,
+    _env: Env,
+    _info: MessageInfo,
+    msg: InstantiateMsg,
+) -> NeutronResult<Response<NeutronMsg>> {
+    deps.api.debug("WASMDEBUG: instantiate");
+    set_contract_version(deps.storage, CONTRACT_NAME, CONTRACT_VERSION)?;
+
+    // TODO: validations
+    CLOCK_ADDRESS.save(deps.storage, &Addr::unchecked(msg.clock_address))?;
+    CONTRACT_STATE.save(deps.storage, &ContractState::Instantiated)?;
+    LP_POSITION.save(deps.storage, &msg.lp_position)?;
+    HOLDER_ADDRESS.save(deps.storage, &msg.holder_address)?;
+
+    Ok(Response::default())
+}
+
+#[entry_point]
+pub fn execute(
+    mut deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    msg: ExecuteMsg,
+) -> NeutronResult<Response<NeutronMsg>> {
+    deps.api
+        .debug(format!("WASMDEBUG: execute: received msg: {:?}", msg).as_str());
+    match msg {
+        ExecuteMsg::Tick {} => try_tick(deps, env, info),
+        ExecuteMsg::WithdrawRewards {} => try_withdraw(deps, env, info),
+    }
+}
+
+
+fn try_tick(deps: DepsMut, env: Env, info: MessageInfo) -> NeutronResult<Response<NeutronMsg>> {
+    let current_state = CONTRACT_STATE.load(deps.storage)?;
+
+    match current_state {
+        ContractState::Instantiated => try_enter_lp_position(deps, env, info),
+        ContractState::LpPositionEntered => no_op(),
+        ContractState::LpPositionExited => no_op(),
+        ContractState::WithdrawComplete => no_op(),
+    }
+}
+
+fn no_op() -> NeutronResult<Response<NeutronMsg>> {
+    Ok(Response::default())
+}
+
+fn try_enter_lp_position(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo, 
+) -> NeutronResult<Response<NeutronMsg>> {
+    let pool_address = LP_POSITION.load(deps.storage)?;
+
+    // get balances of uatom and statom
+    let balances: Vec<Coin> = deps.querier.query_all_balances(env.contract.address)?
+        .into_iter()
+        .filter(|coin| coin.denom == "uatom" || coin.denom == "statom")
+        .collect();
+    
+    // generate astroport Assets from balances
+    let assets: Vec<Asset> = balances.clone().into_iter()
+        .map(|bal| Asset {
+            info: AssetInfo::NativeToken { 
+                denom: bal.denom,
+            },
+            amount: bal.amount,
+        })
+        .collect();
+
+    let provide_liquidity_msg = ProvideLiquidity { 
```

Comment:
This logic is missing a step. You need to first simulate the transaction to know what ratio of ATOM and stATOM you can provide as liquidity. If you provide liquidity in the wrong ratio it will fail.

This is the [query you need to run](https://docs.astroport.fi/docs/develop/smart-contracts/swap-pairs/stableswap-pools#simulation). You can also read more about it here: https://blog.astroport.fi/post/tutorial-how-to-create-a-pair-and-provide-liquidity-on-astroport-using-terra-js

After you get the correct ratio, you can provide the same amount of liquidity. This means that there is a possibility of some left over tokens on one side of the pair. We will then have to do another provide_liquidity call that is single sided (one asset amount is set to 0), which does not require the ratio to match the pool ratio

Additionally, we need to add some tests to test whether our logic works with various ratios of stATOM/ATOM in the pool